### PR TITLE
Tutorial-1

### PR DESCRIPTION
**Pull request by anushreeSonak**

### CODE REVIEW

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
+package stepDefination;
+
+//import static APIAssignment.Product.url;
+
```
Comment (waghmaresuyog) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: remove unwanted  import

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import APIAssignments.ConfigReader;
+//import APIAssignments.Product;
```
Comment (waghmaresuyog) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: same here

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import io.cucumber.java.en.And;
+import io.cucumber.java.en.Given;
+import io.cucumber.java.en.Then;
+import io.cucumber.java.en.When;
+import io.restassured.RestAssured;
+import io.restassured.path.json.JsonPath;
+import io.restassured.response.Response;
+import org.apache.log4j.PropertyConfigurator;
+import org.testng.Assert;
+
```
Comment (waghmaresuyog) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java:15]: remove extra space

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
+
```
Comment (waghmaresuyog) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: same here

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public ProductBDD(String productId, String productName, String productPrice, String productBrand) {
+        this.productId = productId;
+        this.productName = productName;
+        this.productPrice = productPrice;
+        this.productBrand = productBrand;
+    }
+
+    private static Logger logger = Logger.getLogger("ProductBDD.class");
```
Comment (waghmaresuyog) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: declare at  starting of class

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Given("valid page number is available")
+    public void valid_page_number_is_available() {
+        PropertyConfigurator.configure("log4j2.properties");
+    }
+
+    @When("send request to get ProductList")
```
Comment (waghmaresuyog) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: follow naming convention

**Review by waghmaresuyog — CHANGES_REQUESTED** [review]

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 1 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 2 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 3 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java:15]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 4 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 5 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 6 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/pom.xml [hunk 7]
```diff
@@ -16,21 +16,117 @@
...
+        <dependency>
+            <groupId>org.apache.poi</groupId>
+            <artifactId>poi</artifactId>
+            <version>5.1.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.poi</groupId>
+            <artifactId>poi-ooxml</artifactId>
+            <version>5.1.0</version>
+        </dependency>
```
Comment (waghmaresuyog) [inline APIAssignment/pom.xml:43]: remove duplicate dependency

File: APIAssignment/pom.xml [hunk 8]
```diff
@@ -16,21 +16,117 @@
...
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+            <version>2.14.1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <version>2.14.1</version>
+        </dependency>
```
Comment (waghmaresuyog) [inline APIAssignment/pom.xml:53]: same here

File: APIAssignment/pom.xml [hunk 9]
```diff
@@ -16,21 +16,117 @@
...
+            <artifactId>json-path</artifactId>
+            <version>5.3.0</version>
+            <scope>test</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/io.rest-assured/xml-path -->
+        <dependency>
+            <groupId>io.rest-assured</groupId>
+            <artifactId>xml-path</artifactId>
+            <version>5.3.0</version>
+        </dependency>
```
Comment (waghmaresuyog) [inline APIAssignment/pom.xml:73]: Remove duplicate dependency

**Review by waghmaresuyog — CHANGES_REQUESTED** [review]

**Review by waghmaresuyog — APPROVED** [review]
Approving PR

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    private String productBrand;
+
+    private static Logger logger = Logger.getLogger("ProductBDD.class");
+
+    public ProductBDD() {
+        try {
+            url = ConfigReader.getUrl();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
```
Comment (prashantkawa) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: use before method to read URL from properties file

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        this.productName = productName;
+        this.productPrice = productPrice;
+        this.productBrand = productBrand;
+    }
+
+
+    @Given("valid page number is available")
+    public void validatePageNumber() {
+        PropertyConfigurator.configure("log4j2.properties");
+    }
```
Comment (prashantkawa) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: step name and functionality does not match

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                    Assert.assertEquals(mockProduct.productPrice, price);
+                    Assert.assertEquals(mockProduct.productBrand, brand);
+                }
+            });
+        }
+    }
+
+    @And("validate length")
+    public void validateLength() {
+        var response = given().when().get(url).then().extract().asString();
```
Comment (prashantkawa) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: extract response once store it and use wherever needed

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+import static io.restassured.RestAssured.given;
+
+public class ProductBDD {
+    private static String url;
+    private String productId;
+    private String productName;
+    private String productPrice;
+    private String productBrand;
+
+    private static Logger logger = Logger.getLogger("ProductBDD.class");
```
Comment (prashantkawa) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: use final here

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+import io.restassured.RestAssured;
+import io.restassured.path.json.JsonPath;
+import io.restassured.response.Response;
+import org.apache.log4j.PropertyConfigurator;
+import org.testng.Assert;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Logger;
```
Comment (prashantkawa) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: use Logger from Log4j library

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Then("validate content")
+    public void validateContent() {
+        Response response = RestAssured.get(url).then().extract().response();
+        JsonPath jsonObject = new JsonPath(response.asString());
+        List<Object> productList = jsonObject.getList("products");
+        List<ProductBDD> mockData = new ArrayList<>();
+        ProductBDD productLists = new ProductBDD("1", "Blue Top", "Rs. 500", "Polo");
+        mockData.add(productLists);
+        productLists = new ProductBDD("2", "Men Tshirt", "Rs. 400", "H&M");
```
Comment (prashantkawa) [inline APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Try reading this data from file

**Review by prashantkawa — CHANGES_REQUESTED** [review]

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 14 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 13 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 12 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: DONE!

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 11 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: DONE!

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 10 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: DONE !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 12 above)
Comment (prashantkawa) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Not resolved

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 13 above)
Comment (prashantkawa) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Not resolved

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 15 above)
Comment (prashantkawa) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Not resolved

Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 13 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !!

File: APIAssignment/src/test/java/stepDefination/ProductBDD.java (hunk 12 above)
Comment (anushreeSonak) [reply APIAssignment/src/test/java/stepDefination/ProductBDD.java]: Done !

**Review by prashantkawa — APPROVED** [review]
Approving PR

**Pull request merged**